### ISSUE TITLE
Class field initialization order different between target ES2021 and ES2022

### ISSUE THREAD

**Issue by DillonJettCallis** [microsoft/TypeScript#52331]
# Bug Report

### 🔎 Search Terms

ES2022 class field initialization order

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=9#code/JYOwLgpgTgZghgYwgAgApynAtsg3gKGWTgC5kQBXLAI2gG58BfffBAGzgGdPkAVCTmDyFkAelHIEAexCCoFBGClRkYABZwhnCGB5xkMYBDYATEdNlh5i5QAoADlGAA3TSnsZsZdJiwBKAiJmEXFkCm1VDSFDYxNVKWRtITgQKXVoAyNTEX0AXkjgTgA6D18iuAZgi0FVZHyQCAB3PgEwW1xSZABGAAZGPwZqqTYIIrYpAHNbMHKB-CA)

### 💻 Code

```ts
interface Param {
  a: number;
}

class Test {
  // constructor that sets a field
  constructor(private param: Param){
  }

  // use that field to set another field
  a = this.param.a;
}

const t = new Test({a: 10});
console.log(t.a);
```

### 🙁 Actual behavior

If the compiler is set to target "ES2021" then the output will be:

```ts
"use strict";
class Test {
    constructor(param) {
        this.param = param;
        this.a = this.param.a;
    }
}
const t = new Test({ a: 10 });
console.log(t.a);
```

If the target is set to "ES2022" then the output will be:

```ts
"use strict";
class Test {
    param;
    constructor(param) {
        this.param = param;
    }
    a = this.param.a;
}
const t = new Test({ a: 10 });
console.log(t.a);
```

These look logically the same, just utilizing the field syntax in Javascript, but in Javascript fields are initialized BEFORE the constructor is run.

So that means that ES2021 will work, setting the `param` field on `Test` and then use it to set the `a` field, but with ES2022 it will attempt to set `a` first, using the `param` field which has not been initialized yet, thus throwing an error at runtime.

There is no warning or error that this will happen, and I could not find any kind of compiler flag to enable such an error. Even "strictPropertyInitialization" didn't catch this like I thought it might.

### 🙂 Expected behavior

In order to maintain Javascript semantics, I believe that Typescript should not allow field initializers to access these automatic field setting constructor parameters. Typescript should not allow you to do something that will simply not work.

Alternatively, the compiler could be changed to insert the initializer at the end of the constructor. This would maintain existing Typescript behavior at the expense of being different from Javascript.

**Comment by whzx5byb** [comment]
Related: https://github.com/microsoft/TypeScript/issues/50971

**Comment by RyanCavanaugh** [comment]
> Alternatively, the compiler could be changed to insert the initializer at the end of the constructor

I don't think we can change the emit here, but giving a correct error under `strictPropertyInitialization` is definitely desirable here

**Comment by kirbysayshi** [comment]
I'm not sure if this also falls under `strictPropertyInitialization` or not:

```ts
class API {
  private _state = this.initialState();

  constructor(private storage: Map<string, unknown>) {}

  initialState() {
    return this.storage.get("anything");
  }
}
```

This is arguably, to me at least, an example where the `_state` field is not being strictly initialized because it's effectively impossible under ES2022 (!). `this.storage` will not be set by the time `this.initialState()` is executed as a field. It's hard to know right now how many instances of code like this is lurking in a codebase!

[Playground Link](https://www.typescriptlang.org/play?useDefineForClassFields=true&target=9&module=7#code/MYGwhgzhAECCAKBJaBvAUNaAHATgSwDcwAXAU2gH0JiTyBeaYgCzwgDo8A7PYvMEAMo0yACgCUAbjQZowAPadqOAK7BicnCNyFa0ahrABzUgC5oAWTBYAPEq6GANNGWcA1pzkB3TgD4xqAF9pTC4ePkFhUnFUGUwcUmJlHE5GFnZ9HCNSNmNiEQAiME4AT2Z7fMkZIKCgA)

**Comment by RyanCavanaugh** [comment]
It's really unclear how you'd detect that except by checking the method bodies *n* times where at each *n = k* you pretend that only the first `k` class properties have been initialized. But even that doesn't really have a good way to account for what happens in base or derived methods.

**Comment by kirbysayshi** [comment]
Maybe it's better as a lint rule 🤷 ? False positives could at least be ignored then. Something about calling methods from class field properties that reference constructor shorthand-defined properties...

It's really too bad that this is how the ES2022 spec defines the execution order, as it seems to break a lot of existing TS code. For now our workaround will be `"useDefineForClassFields": false`, but this means that code (and our devs' mental models) will drift further and further from how ES actually executes!

**Comment by matthewjamesadam** [comment]
Our team is trying to update to `"target": "ES2022"` and we are running into this issue.  I am not 100% sure I understand the subtleties here, and I definitely don't understand the 'why' of all of this, but I would appreciate any advice on how to move forward.

We have hundreds of classes that use this pattern, with particularly complicated types and constructors:

```
class SomeClass {
  constructor(private context: Context) {}

  // Note that the type for the return value of otherSubElement() may be complicated
  readonly someField = new OtherObject<SomeArg, OtherArg, YetAnotherArg>(context).subElement().otherSubElement();
};
```

all pretty clean.

With es2022 I believe we now need to:

```
class SomeClass {
  constructor(context: Context) {
    this.someObject = new OtherObject<SomeArg, OtherArg, YetAnotherArg>(context).subElement().otherSubElement();
  }

  readonly someField;
};
```

This seems duplicative, and having to migrate all our code to this new syntax will be a chore.  Is there any way around this, or any plan to allow the old syntax to continue being used?  Failing that, as asked above, has anyone found an automated migration tool to make this easier?

Thanks!

**Comment by jakebailey** [comment]
That is expected. TypeScript settled on the "wrong" semantics ("Set") before this feature was spec'd out (as "Define"), which then does not play well with another custom feature (parameter properties) and this is the consequence.

I don't think we have a plan to keep the old one around; it's non-standard. Moving to the official semantics is going to be the thing to do going forward.

Related would be https://github.com/microsoft/vscode/issues/243049, where some tooling was used to do this same sort of thing for VS Code, which also heavily used parameter properties.
